AV-1361: Dataset Spatial Search

dvainio opened the pull request:
Add spatial search to the main page of the dataset list view. The user should be able to either manually draw a bounding box on a map or search by municipality name from a drop down menu of Finnish municipalities.

Implemented in this PR:

- Added new model for storing all bounding box data for all Finnish municipalities
- Added static JSON file for reading in the bounding box data to the table with a command
- Implemented API endpoint for fetching all the municipalities and their bounding boxes in a format which is understood by the select2 drop down
- Implement command for init of the new database table
- Implement command for populating the new table with data
- Add the spatial search to the front end, override the html and css
- Implement the drop down search functionality from scratch
- Implement necessary javascript for text based spatial search
- Implement script for getting bbox data from external service and generate a JSON file

dvainio committed:
AV-1361: Implement bounding box model

Create model.py to ytp_main as a collection for our own custom
models that are not tied to a specific module but has a more
general purpose.

Implement model for storing bounding box data into the database.

Implement function for initializing the models in model.py.

Refs AV-1361

dvainio committed:
AV-1361: Implement click command for ytp model

Implement click command for building the models in ytp_main's
model file.

Implement a click command populating the bounding box model
with data. The data has been pre-harvested and stored in a static
JSON file.

Implement action for storing the bbox data into the model.

Run the commands by adding them to init_ckan.sh

Refs AV-1361

dvainio committed:
AV-1361: Implement endpoint and plugin configuration

Implement an endpoint for fetching all of the bounding boxes
for all municipalities.

Expose the the endpoint by using blueprint in the plugin.

Expose the action for populating the bbox data in the plugin.

Refs AV-1361

dvainio committed:
AV-1361: Implement front-end functionality

Implement snippets for the spatial query by overriding the snippets
and the css from the spatial search module.

Implement own snippet for searching the municipalities through a drop down
menu with select2.

Implement javscript which hooks into the search of the spatial module and performs
a search based on the drop down bounding box data.

Refs AV-1361

dvainio committed:
AV-1361: Add the snippet to the search template

Refs AV-1361

Zharktas reviewed (approved):
It should be documented where the static coordinate file originated if it ever needs an update.

dvainio committed:
AV-1361: Get bbox data script

Script for generating the JSON file which is used by the spatial
data functionality. The script uses the GeoPy client for using the
geocoding service. The script uses the Nominatim service by default,
it has not been tested with anything else. Changing service should
be possible by modifying the code.

Refs AV-1361
